File: algorithms/java/2_add_two_numbers/AddTwoNumbers.java
class AddTwoNumbers {
	public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
		if (l1 == null) {
			return l2;
		}
		if (l2 == null) {
			return l1;
		}
		ListNode dummy = new ListNode(0);
		ListNode cur = dummy;
		int carry = 0;
		while (l1 != null || l2 != null) {
			int val1 = l1 == null ? 0 : l1.val;
			int val2 = l2 == null ? 0 : l2.val;
			int sum = val1 + val2 + carry;
			cur.next = new ListNode(sum % 10);
			carry = sum / 10;
			if (l1 != null) {
				l1 = l1.next;
			}
			if (l2 != null) {
				l2 = l2.next;
			}
			cur = cur.next;
		}
		if (carry != 0) {
			cur.next = new ListNode(carry);
		}
		return dummy.next;
	}
}

File: algorithms/cpp/2_add_two_numbers/AddTwoNumbers.cpp
class AddTwoNumbers {
public:
	ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
		ListNode dummy(0);
		ListNode* prev = &dummy;
		int carry = 0;
		while (l1 != nullptr || l2 != nullptr) {
			int val1 = l1 == nullptr ? 0 : l1->val;
			int val2 = l2 == nullptr ? 0 : l2->val;
			int sum = carry + val1 + val2;
			prev->next = new ListNode(sum % 10);
			carry = sum / 10;
			if (l1 != nullptr) {
				l1 = l1->next;
			}
			if (l2 != nullptr) {
				l2 = l2->next;
			}
			prev = prev->next;
		}
		if (carry != 0) {
			prev->next = new ListNode(carry);
		}
		return dummy.next;
	}
};
